django/db/migrations/migration.py

```diff
@@ -108,20 +121,39 @@ def apply(self, project_state, schema_editor, collect_sql=False):
                 schema_editor.collected_sql.append("-- %s" % operation.describe())
                 schema_editor.collected_sql.append("--")
                 if not operation.reduces_to_sql:
-                    continue
-            # Save the state before the operation has run
-            old_state = project_state.clone()
-            operation.state_forwards(self.app_label, project_state)
+                    return (from_state, to_state)
+            operation.state_forwards(self.app_label, to_state)
             # Run the operation
             atomic_operation = operation.atomic or (self.atomic and operation.atomic is not False)
             if not schema_editor.atomic_migration and atomic_operation:
                 # Force a transaction on a non-transactional-DDL backend or an
                 # atomic operation inside a non-atomic migration.
                 with atomic(schema_editor.connection.alias):
-                    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
+                    operation.database_forwards(self.app_label, schema_editor, from_state, to_state)
             else:
                 # Normal behaviour
-                operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
+                operation.database_forwards(self.app_label, schema_editor, from_state, to_state)
+            return (from_state, to_state)
+
+        def _apply_operations(operations):
```

Comment:
Could these functions be defined as methods instead?

It feels like it would make the execution flow easier to follow and be more in line with the usual coding style; _private_ methods are usually defined instead of relying on internal functions.

Reply:
I made them internal functions to reduce the number of arguments to pass to the function, but I think I can make them private methods yes, I'll give it a shot.

Reply:
Do you like the change? If so, I'll do the same to `unapply`.

Reply:
I just did, let me know what you think.